### ISSUE TITLE
Low hanging fruit of optimization, choice point reuse

### ISSUE THREAD

**Issue by jburse** [jburse/dogelog-moon#35]
A low hanging fruit of optimization is reusing the choice 
point object. We can do this in solve2() and solve_alt()
handling. It comes with no extra cost, since a choice 

point is anyway what sets the stage for these internal
JavaScript calls. We could give solve2() and solve_alt()
an additional parameter, populating it with null in the

first call. And then offering the choice point for reuse.

**Comment by jburse** [comment]
We implemented the change, but have now more question marks
than before. Namely we found:

- There was a bug in solve_alt, we didn't store the trail in mark,
   so the cut variable was bound, but this binding was not undone.

- We might set the choice.tail etc.. to null, to help the garbage
   collector, but there is doubt that this would have some positive impact.

- When we are at it we could also introduce a cutter, which would
  be needed for `setup_call_cleanup/3`.
